library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
+        return manager.exec_module()
 
-        return target_volume
+    def get_manager(self, target):
+        if target == 'remote':
+            return RemoteManager(self.client)
+        if target == 'local':
+            return LocalManager(self.client)
 
-    def delete_volume(self, volume):
-        sleep_interval = 0.25
 
-        # Check the target volume exists and inactive
-        if volume not in self.get_inactive_volumes():
-            return False
+class BaseManager(object):
+    def __init__(self, client):
+        self.client = client
+        self.have = None
+        self.want = Parameters()
+        self.want.client = self.client
+        self.want.update(self.client.module.params)
+        self.changes = Parameters()
 
-        self.api.System.SoftwareManagement.delete_volume(volume)
+    def exec_module(self):
+        changed = False
+        result = dict()
+        state = self.want.state
 
-        while True:
-            time.sleep(sleep_interval)
-            if volume not in self.get_inactive_volumes():
-                break
+        try:
+            if state == "activated":
+                changed = self.activated()
+            elif state == "installed":
+                changed = self.installed()
+            elif state == "present":
+                changed = self.present()
+            elif state == "absent":
+                changed = self.absent()
+        except iControlUnexpectedHTTPError as e:
+            raise F5ModuleError(str(e))
+
+        changes = self.changes.to_return()
+        result.update(**changes)
+        result.update(dict(changed=changed))
+        return result
 
-        return True
+    def _set_changed_options(self):
+        changed = {}
+        for key in Parameters.returnables:
+            if getattr(self.want, key) is not None:
+                changed[key] = getattr(self.want, key)
+        if changed:
+            self.changes = Parameters(changed)
+
+    def _update_changed_options(self):
+        changed = {}
+        for key in Parameters.updatables:
+            if getattr(self.want, key) is not None:
+                attr1 = getattr(self.want, key)
+                attr2 = getattr(self.have, key)
+                if attr1 != attr2:
+                    changed[key] = attr1
+        if changed:
+            self.changes = Parameters(changed)
+            return True
+        return False
 
-    def upload(self, filename):
-        done = False
-        first = True
+    def activated(self):
+        if self.is_activated():
+            return False
+        elif self.is_installed():
+            return self.activate()
+        else:
+            self.install_volume()
+            self.wait_for_device_reboot()
+            return True
 
-        with open(filename, 'rb') as fh:
-            remote_path = "/shared/images/%s" % os.path.basename(filename)
+    def is_activated(self):
+        result = self.software_on_volume()
+        if result:
```

Comment:
i would test the negation here, return none if it succeeds, and then backspace the code below it.

if not result:
    return False

self.have.....